File: ruby/binary_search_tree.rb
require 'bst_node'

class BinarySearchTree

  attr_reader :root

  def initialize
    @root = nil
  end

  def find(value, node = @root)
    # base cases to escape recursive calls
    return nil if node.nil?
    return node if node.value == value

    # use of recursion to navigate through structure, always going to the
    # left child node if target is less than current node value and always
    # right if target is greater than current node value
    if value < node.value
      find(value, node.left)
    elsif value > node.value
      find(value, node.right)
    end
  end

  def insert(value)
    # setting the root here in case it is still nil
    @root = insert_node(@root, value)
  end

  def delete(value)
    # setting the root here in case it is still nil
    @root = delete_node(@root, value)
  end

  def maximum(node = @root)
    if node.right
      maximum_node = maximum(node.right)
    else
      maximum_node = node
    end
    maximum_node
  end

  def depth(node = @root)
    if node.nil?
      return -1;
    else
      left_depth = depth(node.left)
      right_depth = depth(node.right)

      if left_depth > right_depth
        return left_depth + 1
      else
        return right_depth + 1
      end
    end
  end

  def is_balanced?(node = @root)
    return true if node.nil?

    balanced = true
    left_depth = depth(node.left)
    right_depth = depth(node.right)
    balanced = false if (left_depth - right_depth).abs > 1

    if balanced && is_balanced?(node.left) && is_balanced?(node.right)
      return true
    end

    false
  end

  def in_order_traversal(node = @root, arr = [])
    # left children, itself, right children
    if node.left
      in_order_traversal(node.left, arr)
    end

    arr.push(node.value)

    if node.right
      in_order_traversal(node.right, arr)
    end

    arr
  end

  private

  def insert_node(node, value)
    return BSTNode.new(value) if node.nil?

    if (value <= node.value)
      node.left = insert_node(node.left, value)
    elsif (value > node.value)
      node.right = insert_node(node.right, value)
    end

    node
  end

  def delete_node(node, value)
    # if correct node is found, then call remove on it.
    # otherwise, recurse until you find it
    if (value == node.value)
      node = remove(node)
    elsif (value <= node.value)
      node.left = delete_node(node.left, value)
    else (value > node.value)
      node.right = delete_node(node.right, value)
    end
    node
  end

  def remove(node)
    if node.left.nil? && node.right.nil?
      # when node doesn't have children,
      # simply remove it.
      node = nil
    elsif node.left && node.right.nil?
      # when node only has one child,
      # delete it and promote its child to take its place.
      node = node.left
    elsif node.left.nil? && node.right
      node = node.right
    else
      # if node has two children,
      # promote the maximum node in its left subtree to replace itself
      # if that specific node that was promoted has children,
      # then promote
      node = replace_parent(node)
    end
    node
  end

  def replace_parent(node)
    replacement_node = maximum(node.left)
    promote_child(node.left) if replacement_node.left

    # since the replacement_node is still pointing to its old children,
    # we must update its children pointers
    replacement_node.left = node.left
    replacement_node.right = node.right

    replacement_node
  end

  def promote_child(node)
    if node.right
      current_parent = node
      maximum_node = maximum(node.right)
    else
      return node
    end

    current_parent.right = maximum_node.left
  end
end

File: javascript/binary_search_tree.js
//Node
class BSTNode {
  constructor(value){
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

//Tree
class BinarySearchTree {
  constrctor(){
    this.root = null;
  }

  find(value, node = this.root) {
    if(!node) {
      return null;
    } else if (node.value === value) {
      return node;
    } else {
      if(value < node.value) {
        return this.find(value, node.left);
      } else if (value > node.value) {
        return this.find(value, node.right);
      }
    }
  }

  insert(value) {
    this.root = this._insertNode(this.root,value);
    return this;
  }

  delete(value) {
    this.root = this._deleteNode(this.root,value);
    return this;
  }

  maximum(node = this.root) {
    let maximumNode;

    if (node.right !== null) {
      maximumNode = this.maximum(node.right);
    } else {
      maximumNode = node;
    }

    return maximumNode;
  }

  depth(node = this.root) {
    if (!node) {
      return 0
    } else {
      let leftDepth= this.depth(node.left);
      let rightDepth= this.depth(node.right);

      if (leftDepth > rightDepth) {
        return leftDepth + 1;
      } else {
        return rightDepth + 1;
      }
    }
  }

  isBalanced(node = this.root) {
    if (!node) {
      return true;
    }

    let balanced = true;
    let leftDepth = this.depth(node.left);
    let rightDepth = this.depth(node.right);

    if (Math.abs(leftDepth - rightDepth) > 1) {
      balanced = false
    }

    return balanced && this.isBalanced(node.left) && this.isBalanced(node.right)
  }

  inOrderTraversal(node = this.root, arr = []) {
    if (node.left) {
      this.inOrderTraversal(node.left, arr)
    }

    arr.push(node.value);

    if (node.right) {
      this.inOrderTraversal(node.right, arr);
    }

    return arr;
  }

  //methods meant to be private

  _insertNode(node, value) {
    if(!node) {
      return new BSTNode(value);
    }

    if(value <= node.value) {
      node.left = this._insertNode(node.left, value);
    } else if(value > node.value) {
      node.right = this._insertNode(node.right, value);
    }

    return node
  }

  _deleteNode(node, value) {
    if (value == node.value) {
      node = this._remove(node);
    } else if (value <= node.value) {
      node.left = this._deleteNode(node.left, value);
    } else if (value > node.value) {
      node.right = this._deleteNode(node.right, value);
    }

    return node
  }

  _remove(node) {
    if (node.left === null && node.right === null) {
      node = null
    } else if (node.left !== null && node.right === null) {
      node = node.left;
    } else if (node.left === null && node.right !== null) {
      node = node.right;
    } else {
      node = this._replaceParent(node)
    }

    return node;
  }

  _replaceParent(node) {
    let replacementNode = this.maximum(node.left);

    if(replacementNode.left !== null) {
      this._promoteChild(node.left);
    }

    replacementNode.left = node.left;
    replacementNode.right = node.right;

    return replacementNode;
  }

  _promoteChild(node) {
    if (node.right !== null) {
      let currentParent = node;
      let maximumNode = this.maximum(node.right);
    } else {
      return node;
    }

    currentParent.right = maximumNode.left
  }
}
